data/appRateLimiter.go

```diff
@@ -0,0 +1,215 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package data
+
+import (
+	"encoding/binary"
+	"math"
+	"sync/atomic"
+	"time"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/protocol"
+	"golang.org/x/crypto/blake2b"
+
+	"github.com/algorand/go-deadlock"
+)
+
+// appRateLimiter implements a sliding window counter rate limiter for applications
+type appRateLimiter struct {
+	maxSize           uint64
+	serviceRate       uint64
+	serviceRateWindow time.Duration
+
+	seed uint64
+
+	buckets [128]map[crypto.Digest]*appRateLimiterEntry
+	mus     [128]deadlock.RWMutex
+}
+
+func makeAppRateLimiter(maxSize uint64, serviceRate uint64, serviceRateWindow time.Duration) *appRateLimiter {
+	r := &appRateLimiter{
+		maxSize:           maxSize,
+		serviceRate:       serviceRate,
+		serviceRateWindow: serviceRateWindow,
+		seed:              crypto.RandUint64(),
+	}
+	for i := range r.buckets {
+		r.buckets[i] = make(map[crypto.Digest]*appRateLimiterEntry)
+	}
+	return r
+}
+
+type appRateLimiterEntry struct {
+	prev     atomic.Uint64
+	cur      atomic.Uint64
+	interval atomic.Int64 // numeric representation of the current interval value
+}
+
+func (r *appRateLimiter) entry(b int, key crypto.Digest, curInt int64) (*appRateLimiterEntry, bool) {
+	r.mus[b].Lock()
+	defer r.mus[b].Unlock()
+
+	if len(r.buckets[b]) >= int(r.maxSize) {
+		// evict the oldest entry
+		// TODO: evict 10% oldest entries?
+		var oldestKey crypto.Digest
+		var oldestInterval int64 = math.MaxInt64
+		for k, v := range r.buckets[b] {
+			interval := v.interval.Load()
+			if interval < oldestInterval {
+				oldestKey = k
+				oldestInterval = interval
+			}
+		}
```

Comment:
Ring buffers holding the bucket keys in entry order could avoid this range search if you want to spend the x megs of memory on it.

Reply:
used LRU lists

Reply:
i saw, nice work pavel! reading it now